src/rust/engine/fs/store/src/lib.rs

```diff
@@ -1278,44 +1281,33 @@ impl Store {
     destination: PathBuf,
     is_root: bool,
     force_mutable: bool,
+    can_hardlink: bool,
     parent_to_child: &'a HashMap<PathBuf, Vec<directory::Entry>>,
     mutable_paths: &'a BTreeSet<PathBuf>,
-    immutable_inputs: Option<&'a ImmutableInputs>,
     perms: Permissions,
   ) -> BoxFuture<'a, Result<(), StoreError>> {
     let store = self.clone();
     async move {
-      let destination2 = destination.clone();
-      store
-        .local
-        .executor()
-        .spawn_blocking(
-          move || {
-            if is_root {
-              fs::safe_create_dir_all(&destination2)
-            } else {
-              fs::safe_create_dir(&destination2)
-            }
-          },
-          |e| Err(format!("Directory creation task failed: {e}")),
-        )
-        .map_err(|e| {
-          format!(
-            "Failed to create directory {}: {}",
-            destination.display(),
-            e
+      if !is_root {
+        let destination2 = destination.clone();
+        store
+          .local
+          .executor()
+          .spawn_blocking(
+            move || fs::safe_create_dir(&destination2),
+            |e| Err(format!("Directory creation task failed: {e}")),
           )
-        })
-        .await?;
+          .map_err(|e| format!("Failed to create directory {}: {e}", destination.display(),))
+          .await?;
+      }
 
       if let Some(children) = parent_to_child.get(&destination) {
         let mut child_futures = Vec::new();
         for child in children {
           let path = destination.join(child.name().as_ref());
           let store = store.clone();
           child_futures.push(async move {
-            let can_be_immutable =
-              immutable_inputs.is_some() && !mutable_paths.contains(&path) && !force_mutable;
+            let can_be_immutable = !mutable_paths.contains(&path) && !force_mutable && can_hardlink;
```

Comment:
Can skip the lookup if the boolean conditions don't match.
```suggestion
            let can_be_immutable = !force_mutable && can_hardlink && !mutable_paths.contains(&path);
```